### ISSUE TITLE
NoMethodError (undefined method `title' for #<CustomFilter...

### ISSUE THREAD

**Issue by ronilaukkarinen** [mastodon/mastodon#27622]
### Steps to reproduce the problem

1. Upgrade to main version at [a918208](https://github.com/mastodon/mastodon/commit/a918208ec61836f911bd9775d2f2dd6db72ac1a5)
2. Run bundle install && rm -rfv public/packs && yarn install --frozen-lockfile && RAILS_ENV=production bundle exec rails assets:precompile 
3. Restart services

### Expected behaviour

Mastodon should work without errors

### Actual behaviour

Fatal error and "500" notification present

### Detailed description

I noticed it takes time for Explore etc. tabs to populate, but after a while they appear. 

Got a fatal error couple of times (log below).

I haven't seen this before in prior versions, even in 4.3.0-alpha.0. Is it a regression? What causes it? It's related to my muted words in the filter "War, death and destruction". There are normal words there, why does it throw that? 

### Mastodon instance

mementomori.social

### Mastodon version

v4.3.0-alpha.0

### Browser name and version

Google Chrome 118.0.5993.70 (Official Build) (64-bit)

### Operating system

Pop!_OS Linux 22.04 LTS

### Technical details

```
Oct 30 18:11:49 mementomori bundle[786873]: F, [2023-10-30T18:11:49.098377 #786873] FATAL -- : [7d20891b-5303-4486-b404-2fddd2aa2882]
Oct 30 18:11:49 mementomori bundle[786873]: [7d20891b-5303-4486-b404-2fddd2aa2882] NoMethodError (undefined method `title' for #<CustomFilter id: 93, account_id: 109293050593156468, expires_at: nil, phrase: "War, death and destruction", context: ["home", "notifications", "public", "thread", "account"], created_at: "2023-10-19 16:04:59.040990000 +0000", updated_at: "2023-10-19 16:08:48.866924000 +0000", action: "hide">):
```

**Comment by ronilaukkarinen** [comment]
Tested it a bit more. After a while the errors are gone, I think the feeds need some time to populate and before they do, the error appears.

But sometimes the feeds are not populating at all. For example Expore tab gets "Nothing is trending right now. Check back later!"

```
Oct 30 18:48:33 mementomori bundle[827068]: [706bd2c4-ed9b-4d0c-937a-da82c68939fb] NoMethodError (undefined method `title' for #<CustomFilter id: 93, account_id: 109293050593156468, expires_at: nil, phrase: "War, death and destruction", context: ["home", "notifications", "public", "thread", "account"], created_at: "2023-10-19 16:04:59.040990000 +0000", updated_at: "2023-10-19 16:08:48.866924000 +0000", action: "hide">):
Oct 30 18:48:33 mementomori bundle[827068]: [706bd2c4-ed9b-4d0c-937a-da82c68939fb]
Oct 30 18:48:33 mementomori bundle[827068]: [706bd2c4-ed9b-4d0c-937a-da82c68939fb] app/controllers/api/v1/timelines/list_controller.rb:12:in `show'
Oct 30 18:48:33 mementomori bundle[827068]: [706bd2c4-ed9b-4d0c-937a-da82c68939fb] app/controllers/concerns/localized.rb:11:in `set_locale'
Oct 30 18:48:33 mementomori bundle[827068]: [706bd2c4-ed9b-4d0c-937a-da82c68939fb] lib/mastodon/rack_middleware.rb:9:in `call'

Oct 30 18:49:28 mementomori bundle[827089]: F, [2023-10-30T18:49:28.442386 #827089] FATAL -- : [ad3b4770-42b7-425b-8a1f-810382bc9dfa]
Oct 30 18:49:28 mementomori bundle[827089]: [ad3b4770-42b7-425b-8a1f-810382bc9dfa] NoMethodError (undefined method `title' for #<CustomFilter id: 93, account_id: 109293050593156468, expires_at: nil, phrase: "War, death and destruction", context: ["home", "notifications", "public", "thread", "account"], created_at: "2023-10-19 16:04:59.040990000 +0000", updated_at: "2023-10-19 16:08:48.866924000 +0000", action: "hide">):
```

I keep seeing it in Explore tab which is empty every once in a while and then in some profiles.

Will keep my version in the working one for now but I hope this gives you insight...

**Comment by ronilaukkarinen** [comment]
Attempted to reproduce it again with Vanilla Mastodon (main) and succeeded.

1) Make sure in your db you have lots of muted word filters
2) Get latest main and restart server
3) Watch log carefully while browsing: `sudo journalctl -u mastodon-web.service -f `

This is Vanilla Mastodon with latest main, I see some empty feeds like Explore and Lists with my latest db dump:

![image](https://github.com/mastodon/mastodon/assets/1534150/336bf8a1-d220-4bbd-9ba9-1fa7d5bbaf7e)

**Comment by jamesadney** [comment]
I didn't have time to do a bisect and figure out the root cause, but this patch got rid of my 500s. After defining the `title` method I started getting "undefined method" for `filter_action` too.

I don't think `REST::FilterSerializer` or `CustomFilter` have changed recently. It might have something to do with `filter_action` and `title` being attribute aliases? Calling `title` and `filter_action` directly on a filter in the console works. Maybe this is a bug in active record or active model serializers?

```diff
--- a/app/serializers/rest/filter_serializer.rb
+++ b/app/serializers/rest/filter_serializer.rb
@@ -9,6 +9,14 @@ class REST::FilterSerializer < ActiveModel::Serializer
     object.id.to_s
   end
 
+  def filter_action
+    object.action
+  end
+
+  def title
+    object.phrase
+  end
+
   def rules_requested?
     instance_options[:rules_requested]
   end
```

**Comment by ronilaukkarinen** [comment]
Thank you @jamesadney so much! I tested it in local and no errors, so was brave enough to push it to production. Testing it on my production instance this evening but so far so good.

**Comment by vmstan** [comment]
> I didn't have time to do a bisect and figure out the root cause, but this patch got rid of my 500s. After defining the `title` method I started getting "undefined method" for `filter_action` too.
> 
> I don't think `REST::FilterSerializer` or `CustomFilter` have changed recently. It might have something to do with `filter_action` and `title` being attribute aliases? Calling `title` and `filter_action` directly on a filter in the console works. Maybe this is a bug in active record or active model serializers?
> 
> ```diff
> --- a/app/serializers/rest/filter_serializer.rb
> +++ b/app/serializers/rest/filter_serializer.rb
> @@ -9,6 +9,14 @@ class REST::FilterSerializer < ActiveModel::Serializer
>      object.id.to_s
>    end
>  
> +  def filter_action
> +    object.action
> +  end
> +
> +  def title
> +    object.phrase
> +  end
> +
>    def rules_requested?
>      instance_options[:rules_requested]
>    end
> ```

Would you submit as a PR?

**Comment by jamesadney** [comment]
> Would you submit as a PR? @jamesadney

@vmstan I'm reluctant to submit a PR before understanding what's going on here. Unfortunately I haven't been able to narrow down the underlying issue. If I do I'll post a PR for sure, but I don't have a ton of time to investigate.

**Comment by jamesadney** [comment]
I poked around on this a little more. I'm still not sure what the cause is, but I've been able to reproduce it in development with caching enabled (`touch tmp/caching-dev.txt`). If I `rm tmp/caching-dev.txt` and restart the server the error goes away.

**Comment by ronilaukkarinen** [comment]
Still seeing this (main fetched today). Has @renchap any idea?

**Comment by ClearlyClaire** [comment]
This looks like a bug with caching, but I don't understand in which situation it would happen, and I am unable to reproduce it. Do you perhaps have old versions of the code running by accident?

**Comment by ClearlyClaire** [comment]
cc @byroot, who is very knowledgeable about Rails cache and helped debug another weird issue in the past

**Comment by byroot** [comment]
> If you restart Ruby between the two statements (with a cache that is still warm), the read statement will fail with a `NoMethodError`.

Yeah, sounds like loading a record from the cache don't trigger attribute definition. I'm in vacation right now, so I may not get time to work on this before a week or two (dunno really).

You can confirm by suspicion by calling `CustomFilter.define_attribute_methods` before the `cache.read`. If that's indeed it, I can probably fix it on the Rails side.

**Comment by ClearlyClaire** [comment]
@byroot thanks for the prompt reply! Calling `CustomFilter.define_attribute_methods` before the `cache.read` does indeed work!

**Comment by byroot** [comment]
Thanks, I'll fix it in Rails soon-ish then.

**Comment by byroot** [comment]
I'll see about fixing it for the old format though.

**Comment by byroot** [comment]
https://github.com/rails/rails/pull/50594

**Comment by ClearlyClaire** [comment]
Thanks for looking into it! I missed the marshal format update, but we still haven't published a release running Rails 7.1, so it will have to wait a bit…

**Comment by byroot** [comment]
> e still haven't published a release running Rails 7.1, so it will have to wait a bit…

I think you can do this anyway by flushing the cache, see: https://github.com/mastodon/mastodon/pull/28609
